crypto/rand/randfile.c

```diff
@@ -83,26 +83,37 @@ int RAND_load_file(const char *file, long bytes)
         return 0;
 
 #ifndef OPENSSL_NO_POSIX_IO
-    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode)) {
-        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_NOT_A_REGULAR_FILE);
-        ERR_add_error_data(2, "Filename=", file);
-        return -1;
-    }
+    /*
+     * If not a regular file, assume it's a device and don't try to
+     * read everything.
+     */
+    if (stat(file, &sb) >= 0 && !S_ISREG(sb.st_mode) && bytes < 0)
+        bytes = 256 / 8;
 #endif
+
     if ((in = openssl_fopen(file, "rb")) == NULL) {
         RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_CANNOT_OPEN_FILE);
         ERR_add_error_data(2, "Filename=", file);
         return -1;
     }
 
+    /* Read the file, in chunks. */
     for ( ; ; ) {
         if (bytes > 0)
```

Comment:
This isn't a showstopper but I think I'd prefer to see the bytes < 0 check outside the loop:
```
bytes = bytes < 0 ? RAND_FILE_SIZE : bytes;
```

I think the code would be easier to understand this way and it avoids a couple of bytes < 0 checks inside the loop.

Reply:
I don't think you can move it outside the loop, since bytes changes each pass through and you get have a small number of bytes left.

Reply:
How about:
```
bytes = bytes < 0 ? RAND_FILE_SIZE : bytes;
do {
    n = bytes < rand_file_size ? bytes : rand_file_size;
    fread(...)
    ...
    if (i <= 0) break;
    RAND_add(...)
    ret += i
} until ((bytes -= i) <= 0)
```

Apologies for the brevity, typing one handed is doubleplusungood.

Reply:
Anyway, whichever you think looks easier to comprehend...